### ISSUE TITLE
Understand the significance of different errors and their ranges, and few hyperparameters

### ISSUE THREAD

**Issue by shimonymittal** [sintel-dev/Orion#510]
Hi @blaskowitz100,

I ran the code provided by you in comment 'https://github.com/sintel-dev/Orion/issues/143#issuecomment-768640950'.
The code runs fine after rectifying few errors as per my data. But I have 2 questions:
1. I want to know the significance of 'n_iterations_critic'
2. Where in the code of train step have we used epoch to train the model for the specified number of epochs?
3. I capture history while fitting the model, the history results are as below, whereas I was expecting all the loss values in each n_iterations_critic step, instead of one:
history.history.keys() 
dict_keys(['Cx_total', 'Cx_valid', 'Cx_fake', 'Cx_gp_penalty', 'Cz_total', 'Cz_valid', 'Cz_fake', 'Cz_gp_penalty', 'EG_total', 'EG_fake_gen_x', 'EG_fake_gen_z', 'G_rec'])
[0.02138495445251465]
[4.742159366607666]
[-4.880722522735596]
[0.015994776040315628]
[-3.5925278663635254]
[-2.8535757064819336]
[-1.146549105644226]
[0.040759697556495667]
[2.1441867351531982]
[-0.054472602903842926]
[1.587338924407959]
[0.061132051050662994]
4. Also. I don't know whether above vlues are satisfactoy, what values to improve and what to expect in each loss (I did refer issue 'https://github.com/sintel-dev/Orion/issues/441' but I am still not able to comprehend)

_Originally posted by @shimonymittal in https://github.com/sintel-dev/Orion/issues/143#issuecomment-1926638291_

**Comment by sarahmish** [comment]
Hi @shimonymittal, these are all very interesting questions! 

1. the parameter `n_iterations_critic` refers to the number of iterations used to update the critic models. More specifically if `n_iterations_critic=5` then the critic models' weights are updated five times before proceeding to the generator.
2. it's part of the `_fit` function https://github.com/sintel-dev/Orion/blob/b534a1a81404d511425fd56934d8dd442845e48d/orion/primitives/tadgan.py#L308
3. unfortunately, for formatting purposes, we only record the average loss of all iterations. to view all the results, please change the following code and remove the aggregation function https://github.com/sintel-dev/Orion/blob/b534a1a81404d511425fd56934d8dd442845e48d/orion/primitives/tadgan.py#L332-L333 you might need to return the value separately to keep the size of the array consistent
4. for simplicity, I recommend looking at the mse value to check if the results are satisfactory or not (denoted as rec)

**Comment by shimonymittal** [comment]
Hi @sarahmish,
Thankyou for the explanation. So when I try the code for the same datasets you have used in the paper specifically, the code works perfect. But when I am using it for my dataset, the reconstruction is not good at all. I am using same window size for my data to keep it similar but what hyperparameters or addition of layers should I perform to get good results.

I tried training the model for 500 epochs as my training data size is (48520, 100, 1)
Keeping 150 lstm units, 128 cnn filters, 4 cnn blocks
but still no good reconstruction possible.
<img width="703" alt="TrainingReconstructionImage" src="https://github.com/sintel-dev/Orion/assets/95504257/cf9b928e-a44f-4e7e-9e9e-b55daf3b3e38">

**Comment by sarahmish** [comment]
@shimonymittal you will need to tweak the model a little bit to get better reconstruction. I first recommend lowering the number of epochs to 10 or so, see if tweaking the hyperparameters is yielding better results, then try to increase the epochs to somewhere around 35. 

I have another observation here is that your `error` signal (in red) is highly sensitive to the last _k_ sequences (clear spike). If you remove this section from `find_anomalies` you might get better results in the detection.

hope this helps!

**Comment by shimonymittal** [comment]
@sarahmish Thankyou so much for your input. I rectified above error. Took all parameters as per the model here. I am getting rec error value as low as 0.19-0.099. But the reconstruction results are very poor. Maybe this model is not suitable for real application data. Let me know if you could understand anything out of the results for improving the model.
<img width="740" alt="TrainingReconstructionImage_200Epochs_BatchSize128_RestAllParamsSame" src="https://github.com/sintel-dev/Orion/assets/95504257/a5e9e9a2-9644-45b0-9f4f-adb57f2f24bc">
<img width="738" alt="TrainingReconstruction_200Epochs_BatchSize128_RestAllParamsSame_G_recValues" src="https://github.com/sintel-dev/Orion/assets/95504257/cb8a1852-069b-4727-b21a-1eb4a9360ddd">

**Comment by sarahmish** [comment]
Hi @shimonymittal, we have used `tadgan` before for real-world datasets, it works but need patience in exploring the hyperparameters.

I typically always recommend `aer` rather than `tadgan` since it is a faster model with consistently good scores in the benchmark. I would recommend trying it!
